### PR TITLE
Add note about performance of typed containers from the interpreter.

### LINKED ISSUE #6435: Creating coo_sparse matrix from numba typed List is significantly slower than Python list or numpy arrays
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

```python
import numpy as np
from numba.typed import List
from time import time
from scipy.sparse import coo_matrix


N = 1_000_000
M = 1_000_000
i = np.random.randint(0, N, size=M)
j = np.random.randint(0, N, size=M)
v = np.random.rand(M)

# Numpy arrays
t0 = time()
A = coo_matrix((v, (i, j)))
print(f"Numpy array: {time() - t0:.3f}")

# Numba typed List
i, j, v = [List(x) for x in [i, j, v]]

t0 = time()
A = coo_matrix((v, (i, j)))
print(f"Numba typed List: {time() - t0:.3f}")

# Python list
i, j, v = [list(x) for x in [i, j, v]]

t0 = time()
A = coo_matrix((v, (i, j)))
print(f"Python list: {time() - t0:.3f}")

# Numba typed List + conversion time
i, j, v = [List(x) for x in [i, j, v]]

t0 = time()
i, j, v = [np.array(x) for x in [i, j, v]]
A = coo_matrix((v, (i, j)))
print(f"Numba typed List + conversion time: {time() - t0:.3f}")

# Python list + conversion time
i, j, v = [list(x) for x in [i, j, v]]

t0 = time()
i, j, v = [np.array(x) for x in [i, j, v]]
A = coo_matrix((v, (i, j)))
print(f"Python list + conversion time: {time() - t0:.3f}")

```
```bash
Numpy array: 0.001
Numba typed List: 30.124
Python list: 0.305
Numba typed List + conversion time: 14.727
Python list + conversion time: 0.411
```

### LINKED ISSUE #6439: Numba Dict implementation lot slower than pure python Dict Implementation
```
Jitted Function:
    -----------------
    
    from numba.typed import List, Dict

    dict_A = Dict.empty(
             key_type=types.Tuple((types.unicode_type, types.unicode_type, types.unicode_type)),
             value_type=types.int64[:],
         )
    dict_B = Dict.empty(
         key_type=types.Tuple((types.unicode_type, types.unicode_type, types.unicode_type)),
         value_type=types.int64,
    )
    common_keys = List()  # This is list of tuples of type (types.unicode_type, types.unicode_type, types.unicode_type)
    
    @jit(nopython=True)
    def final_calculations(common_keys, dict_A, dict_B):
        final_out_dict = Dict.empty(
            key_type=types.int64,
            value_type=types.int64,
        )
        for key in common_keys:
            for v in dict_A.get(key):
                if key in dict_B:
                    if v not in final_out_dict:
                        final_out_dict[v] = 0
                    final_out_dict[v] += dict_B[key]
        return final_out_dict


       Non Jitted function:
        --------------------
        
        def final_calculations(common_keys, dict_A, dict_B):
            final_out_dict = defaultdict(int)
            for key in common_keys:
                for v in dict_A.get(key):
                    final_out_dict[v] += dict_B[key]
            return final_out_dict
           
```
Running `timeit` on Jitted functions shows it much slower than Non-Jitted function.
```
Results:

Jitted function:

    common_keys = List()
    dict_B[("4tglt", "101", "9")] = 90
    dict_B[("4tglt", "102", "102")] = 30
    common_keys.append(("4tglt", "101", "9"))
    dict_A[("4tglt", "101", "9")] = np.asanyarray([15013958])

%timeit final_calulations(common_keys, dict_A, dict_B)
27.3 µs ± 321 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
```
Non-Jitted Function:

        dict_B = {}
        dict_B[("4tglt", "101", "9")] = 90
        dict_B[("4tglt", "102", "102")] = 30
        common_keys = [("4tglt", "101", "9")]
        dict_A = {}
       dict_A[("4tglt", "101", "9")] = [15013958]

%timeit final_calulations(common_keys, dict_A, dict_B)
610 ns ± 5.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Why is numba `Dict` operations lot slower than pure python Dict operations?

## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title. Also makes a header for another section more clear.

Closes #6439
Closes #6435

### CODE REVIEW

File: docs/source/user/faq.rst [hunk 1]
```diff
@@ -150,6 +150,17 @@ In this case, vectorization is rejected because the vectorized code may behave
 differently.  This is a case to try turning on ``fastmath=True`` to allow
 fastmath instructions.
 
+Why are the ``typed`` containers slower when used from the interpreter?
+-----------------------------------------------------------------------
+
+The Numba ``typed`` containers found in ``numba.typed`` e.g.
+``numba.typed.List`` store their data in an efficient form for access from JIT
+compiled code. When these containers are used from the CPython interpreter, the
```
Comment (EPronovost) [inline docs/source/user/faq.rst:158]: Should this be 2 sentences? i.e. `... from/to the container format. This process is ...`

Comment (stuartarchibald) [reply docs/source/user/faq.rst:158]: Good point :)

Comment (stuartarchibald) [reply docs/source/user/faq.rst:158]: b2f68ba  fixes

**Review by EPronovost — DISMISSED** [review]

**Comment by stuartarchibald** [comment]
Thanks for the review @EPronovost

File: docs/source/user/faq.rst [hunk 2]
```diff
@@ -150,6 +150,16 @@ In this case, vectorization is rejected because the vectorized code may behave
...
+data involved has to be converted from/to the container format. This process is
+relatively costly and as a result impacts performance. In JIT compiled code no
+such penalty exists and so operations on the containers are much quicker and
+often faster than the pure python equivalent.
```
Comment (sk1p) [inline docs/source/user/faq.rst:162]: Nitpick: capitalization of Python

```suggestion
often faster than the pure Python equivalent.
```

Comment (stuartarchibald) [reply docs/source/user/faq.rst]: thanks

**Review by sklam — APPROVED** [review]
LGTM

**Pull request merged**